### PR TITLE
[Relay] Mix mode type inference

### PR DESCRIPTION
**Pull request by lixiaoquan**
@mbrookhart @jroesch Could you please review? Thanks

### CODE REVIEW

**Review by mbrookhart — APPROVED** [review]
LGTM

File: include/tvm/relay/expr_functor.h [hunk 1]
```diff
@@ -408,6 +409,73 @@ Expr PostOrderRewrite(const Expr& expr, ExprRewriter* rewriter);
...
+ * match the dataflow types, or a node who's inputs have all been processed, it visits the current
+ * leaf via fvisit_leaf.
+ *
+ * This function should be used internally to other classes to implement mixed-mode traversals. The
+ * expectation is that fvisit_leaf will perform recursive analysis within mixed-mode traversal if it
+ * hits a non-dataflow node.
+ *
+ * fcheck_visited and fvisit_leaf are templated to encourage compiler inlining.
+ */
+template <typename FCheckVisited, typename FVisitLeaf>
```
Comment (jroesch) [inline include/tvm/relay/expr_functor.h:430]: Why do we need to expose this in the header?

Comment (lixiaoquan) [reply include/tvm/relay/expr_functor.h:430]: It's a template function and we need to use it in type_infer.cc

Comment (mbrookhart) [reply include/tvm/relay/expr_functor.h:430]: InferType directly extends ExprFunctor instead of using ExprVisitor or ExprMutator

**Comment by jroesch** [comment]
@lixiaoquan sorry I forgot to merge this one, going to land now.

**Review by jroesch — APPROVED** [review]

**Pull request merged**